### ISSUE TITLE
support htmlwidgets in RCloud?

### ISSUE THREAD

**Issue by gordonwoodhull** [att/rcloud#1435]
We have some ad hoc support for the RStudio html ecosystem via shared.R and rcloud.shiny, but in a sense we went about this in a backwards way, watching what failed and patching it up.

Now there is htmlwidgets, which seems like it should be a nice standard for embedding HTML/JavaScript based visualizations into cells or rcloud.web applications. 

I wonder how hard it would be to hook in at a deeper level and directly support this infrastructure.

**Comment by gordonwoodhull** [comment]
Related links:

htmlwidgets/Zeppelin: https://github.com/ramnathv/htmlwidgets/issues/146
htmlwidgets/Jupyter: https://github.com/ramnathv/htmlwidgets/issues/144

**Comment by gordonwoodhull** [comment]
Basically it looks like htmlwidgets is currently an rmarkdown-only solution, a standard which is documented on the client side but not the server side. 

(Brings me back to my days of hacking Microsoft OLE, which also documented controls but containers not so well.)

**Comment by gordonwoodhull** [comment]
Of course, we can use rmarkdown and do support rmarkdown cells, so maybe that is a solution here.

**Comment by gordonwoodhull** [comment]
This thread is also relevant: https://github.com/ramnathv/htmlwidgets/issues/86

**Comment by gaborcsardi** [comment]
First working version: https://github.com/MangoTheCat/rcloud/commit/2a8d4342cf26fc77e0b4ad8d0ae67afb7cde1a85

It uses iframes and the srcdoc attribute, which is good and bad. Good, because widgets are totally independent, no JS interference. Bad, because it does not work out of the box on IE. But there is a good polyfill: https://github.com/jugglinmike/srcdoc-polyfill

**Comment by shaneporter** [comment]
Don't think IE support is a requirement, is it @gordonwoodhull?

**Comment by gaborcsardi** [comment]
@shaneporter Nevertheless, the polyfill is really simple, so why not?

**Comment by gordonwoodhull** [comment]
Awesome, @gaborcsardi, please open a PR!

I am a little disappointed over iframes since they don't integrate very well, but let's start there and see if we can improve it over time. I assume that's mostly because of JS library version conflicts?

We don't currently support IE because it has "different" websocket "standards" (like pong messages), so that's not a concern right now.

**Comment by gaborcsardi** [comment]
@gordonwoodhull Regarding iframes, it is mainly because of the simplicity. I can easily pull out a self-contained HTML page from `htmlwidgets`. If it is not an iframe, then I need to parse the HTML, pull out the different parts, pass them to rcloud, etc. It is not too bad, but the parsing is a lot more error-prone, and also potentially sensitive to future changes in `htmlwidgets`.

The current approach is quite robust I think. Plus, as you mention, we also get rid of all JS library conflicts.

IMO iframes integrate relatively well nowadays, although sizing might be an issue, that is not perfect yet. I need to read a bit about how `htmlwidgets` does it. 

But in any case, I can try to get rid of the iframes.

**Comment by gaborcsardi** [comment]
@gordonwoodhull I am trying to make htmlwidgets work in mini.html. I was wondering what kind of API you would like. How about sg. like this:
```r
library(DT)
...
    rcw.set("#mytable", htmlwidget(DT(dataset)))
...
```
and `htmlwidget` would just return the HTML for the widget, in an iframe? 

Actually, as I see it, if I implement an `as.character` method for `htmlwidget` objects, then the conversion is done automatically and this will work:
```r
...
    rcw.set("#mytable", DT(dataset))
...
```

Is this OK?

**Comment by gordonwoodhull** [comment]
That looks great! Do you anticipate any ill side-effects of the `as.character` approach?

Another possibility to consider (depending on time) is to implement a similar behavior to images, so that the user can resize the htmlwidgets inside the notebook or dashboard. @dougmet or @shaneporter can tell you about this functionality; it's used a lot in RCAP.

This would probably require sending the data in its own ocap (see `htdocs/js/session.js`) when it is printed to an output context, so I think it's independent of the character-based approach.

(Presumably exporting images would not work, since htmlwidgets usually depend on the JS side to produce content.)

**Comment by gaborcsardi** [comment]
I think `as.character` is safe, I don't see any problems. 

Just pushed it, it works great in mini.html.

But I really need to do sg about the sizing now, it is probably even more important in mini.html. I know about the stretchy plots, let me do some research on how `htmlwidgets` does this, and then whether I can come up with sg reasonable.

**Comment by gaborcsardi** [comment]
Note to myself: there is a detailed guide for html widget sizing: https://cran.r-project.org/web/packages/htmlwidgets/vignettes/develop_sizing.html

Ideally we would need to capture resizing and pass it on to the widget, but I am not sure if this will be possible with the iframes. Probably not.

If not, then I will try to get rid of the iframes tomorrow.

**Comment by gordonwoodhull** [comment]
Implemented in #2016, still with iframes but they are mostly invisible.
